crates/libcontainer/src/process/memory_policy.rs

```diff
@@ -0,0 +1,542 @@
+use oci_spec::runtime::MemoryPolicyModeType;
+
+use crate::syscall::{Syscall, SyscallError};
+
+#[derive(Debug, thiserror::Error)]
+pub enum MemoryPolicyError {
+    #[error("Invalid memory policy flag: {0}")]
+    InvalidFlag(String),
+
+    #[error("Invalid node specification: {0}")]
+    InvalidNodes(String),
+
+    #[error("Invalid memory policy mode: {0}")]
+    InvalidMode(String),
+
+    #[error("Incompatible flag and mode combination: {0}")]
+    IncompatibleFlagMode(String),
+
+    #[error("Mutually exclusive flags: {0}")]
+    MutuallyExclusiveFlags(String),
+
+    #[error("Syscall error: {0}")]
+    Syscall(#[from] SyscallError),
+}
+
+type Result<T> = std::result::Result<T, MemoryPolicyError>;
+
+pub fn setup_memory_policy(
+    memory_policy: &Option<oci_spec::runtime::LinuxMemoryPolicy>,
+    syscall: &dyn Syscall,
+) -> Result<()> {
+    if let Some(policy) = memory_policy {
+        // Memory policy mode constants from Linux UAPI (include/uapi/linux/mempolicy.h)
+        const MPOL_DEFAULT: i32 = 0;
+        const MPOL_PREFERRED: i32 = 1;
+        const MPOL_BIND: i32 = 2;
+        const MPOL_INTERLEAVE: i32 = 3;
+        const MPOL_LOCAL: i32 = 4;
+        const MPOL_PREFERRED_MANY: i32 = 5;
+        const MPOL_WEIGHTED_INTERLEAVE: i32 = 6;
+
+        // Memory policy flag constants from Linux UAPI
+        const MPOL_F_NUMA_BALANCING: u32 = 1 << 13; // 0x2000
+        const MPOL_F_RELATIVE_NODES: u32 = 1 << 14; // 0x4000
+        const MPOL_F_STATIC_NODES: u32 = 1 << 15; // 0x8000
+
+        let base_mode = match policy.mode() {
+            MemoryPolicyModeType::MpolDefault => MPOL_DEFAULT,
+            MemoryPolicyModeType::MpolPreferred => MPOL_PREFERRED,
+            MemoryPolicyModeType::MpolBind => MPOL_BIND,
+            MemoryPolicyModeType::MpolInterleave => MPOL_INTERLEAVE,
+            MemoryPolicyModeType::MpolLocal => MPOL_LOCAL,
+            MemoryPolicyModeType::MpolPreferredMany => MPOL_PREFERRED_MANY,
+            MemoryPolicyModeType::MpolWeightedInterleave => MPOL_WEIGHTED_INTERLEAVE,
+        };
+
+        let mut flags_value: u32 = 0;
```

Comment:
Personally, I think it would be easier to read if the validation and the construction of flags_value were separated.